spatialos-sdk/src/worker/entity.rs

```diff
@@ -36,36 +45,151 @@ impl Entity {
+
         self.components.insert(
-            C::ID,
-            ComponentData {
-                raw_data,
-                drop_fn: drop_raw::<C>,
+            id,
+            Worker_ComponentData {
+                reserved: ptr::null_mut(),
+                component_id: id,
+                schema_type: ptr::null_mut(),
+                user_handle: component.user_handle,
             },
         );
+
+        Ok(())
     }
 
     pub fn get<C: Component>(&self) -> Option<&C> {
         self.components
             .get(&C::ID)
-            .map(|data| unsafe { &*(data.raw_data.user_handle as *const _) })
+            .map(|data| unsafe { &*(data.user_handle as *const _) })
+    }
+
+    pub(crate) fn raw_component_data(&self) -> RawEntity {
+        RawEntity::new(self.components.values())
+    }
+
+    fn pre_add_check(&self, id: ComponentId) -> Result<(), String> {
+        if self.components.contains_key(&id) {
+            return Err(format!(
+                "Duplicate component with ID {} added to `Entity`.",
+                id
+            ));
+        }
+
+        if !self.database.has_vtable(id) {
+            return Err(format!(
+                "Could not find a vtable implementation for component {}",
+                id
+            ));
+        }
+
+        Ok(())
     }
+}
 
-    pub(crate) fn raw_component_data(&self) -> Vec<Worker_ComponentData> {
-        self.components.values().map(|data| data.raw_data).collect()
+impl Default for Entity {
+    fn default() -> Self {
+        Entity {
+            components: HashMap::new(),
+            database: get_component_database(),
+        }
     }
 }
 
 impl Drop for Entity {
     fn drop(&mut self) {
-        for data in self.components.values() {
-            unsafe {
-                (data.drop_fn)(data.raw_data.user_handle);
-            }
+        for component_data in self.components.values() {
+            let id = component_data.component_id;
+
+            let vtable = self.database.get_vtable(id).unwrap();
+
+            let free_data_func = vtable
+                .component_data_free
+                .unwrap_or_else(|| panic!("No component_data_free method defined for {}", id));
+
+            unsafe { free_data_func(id, ptr::null_mut(), component_data.user_handle) };
         }
     }
 }
 
-unsafe fn drop_raw<T>(raw: *mut std::ffi::c_void) {
-    component::handle_free::<T>(raw);
+// Required for when we call Entity::raw_component_data() and want a Vec<Worker_ComponentData> rather
+// than a Vec<&Worker_ComponentData> which most callers *will* want due to how Worker_Entity is structured.
+pub(crate) struct RawEntity {
+    pub components: Vec<Worker_ComponentData>,
+    database: ComponentDatabase,
+}
+
+impl RawEntity {
+    pub fn new<'a, I>(original_data: I) -> Self
+    where
+        I: Iterator<Item = &'a Worker_ComponentData>,
+    {
+        let database = get_component_database();;
```

Comment:
Double `;`? I'm surprised, I thought that was a compile error 🤔

Reply:
I guess its interpreted as `get_component_database()` and then a no-op?